projects/clr/hipamd/src/hip_graph_internal.cpp

```diff
@@ -418,32 +419,49 @@ hipError_t GraphExec::CaptureAndFormPacketsForGraph() {
     }
 
     if (node->GraphCaptureEnabled()) {
-      // Start of a potential batch - try to capture packets for this node
-      std::vector<uint8_t*> currentBatch;
-      std::vector<std::string> currentKernelNames;
+      // Start of a new batch
+      PacketBatch newBatch;
+      PacketNode* lastNode = nullptr;
 
       // Collect packets from consecutive captured nodes
       size_t j = i;
-      size_t capturedNodeCount = 0;
       while (j < topoOrder_.size() && topoOrder_[j]->GraphCaptureEnabled()) {
         auto& currentNode = topoOrder_[j];
-        status = currentNode->CaptureAndFormPacket(GetKernelArgManager(), &currentBatch,
-                                                             &currentKernelNames);
 
-        if (status != hipSuccess || currentBatch.empty()) {
+        // Capture packets for this node
+        std::vector<uint8_t*> nodePackets;
+        std::vector<std::string> nodeKernelNames;
+        status = currentNode->CaptureAndFormPacket(GetKernelArgManager(), &nodePackets,
+                                                   &nodeKernelNames);
+
+        if (status != hipSuccess || nodePackets.empty()) {
           LogError("Packet capture failed");
           return status;
         }
+
+        // Create PacketNodes for this node's packets
+        std::vector<PacketNode*> nodePacketNodes;
+        for (size_t k = 0; k < nodePackets.size(); ++k) {
+          PacketNode* packetNode = new PacketNode(nodePackets[k], nodeKernelNames[k], 
```

Comment:
Also the created nodes, where are they released. You are putting this on the heap.